### ISSUE TITLE
Union Type silently ignores input in constructor

### ISSUE THREAD

**Issue by sbrother** [samuelcolvin/pydantic#1887]
# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.6
            pydantic compiled: True
                 install path: /home/sam/company/verify/venv/lib/python3.8/site-packages/pydantic
               python version: 3.8.2 (default, Jul 16 2020, 14:00:26)  [GCC 9.3.0]
                     platform: Linux-5.4.0-42-generic-x86_64-with-glibc2.29
     optional deps. installed: []

```
As seen in the snippet below, pydantic seems to interpret an explicit instance of `B` as a default constructed `A` below. This is related to the existing bugs in Union, but is possibly easier to fix since the correct type information for B is passed to the union type in the constructor. It's very unexpected behaviour and leads to silent data loss.
 
```py
from pydantic import BaseModel
from typing import Union, Optional


class A(BaseModel):
    foo: Optional[str]
    same: Optional[str]

class B(BaseModel):
    same: Optional[str]
    bar: Optional[str]

class C(BaseModel):
    baz: Union[A, B]

example = C(baz=B(bar='hello'))
print(example)
```
returns
```
baz=A(foo=None, same=None)
```

**Comment by noelevans** [comment]
If you change class A to this:

```
class A(BaseModel):               
    foo: Optional[str]            
    same: Optional[str]           
                                  
    class Config:                 
        from pydantic import Extra
        extra = Extra.allow       
```

The output becomes 
`baz=A(foo=None, same=None, bar='hello')`

**Comment by PrettyWood** [comment]
Hello @sbrother 
This is a known issue. See #619 and #1423 for more information.
The problem is that it first tries to match with `A` and succeeds as there is no missing value (`foo` and `same` has a default value `None` and `A` ignores extra parameters by default).
You could fix this by setting in `A`
```python
    class Config:
        extra = 'forbid'
```
or by changing the union into `[B, A]`

**Comment by sbrother** [comment]
Thanks @PrettyWood , I figured as much, and `extra = 'forbid'` is how I ended up solving the problem. Like the commenter in #1423 though -- I think this is a fairly important issue since it leads to silent unexpected data loss that would not happen in an alternative like `dataclass` or protocol buffers. 

I understand the issues that arise when checking field types against candidate types, but in this case the data is not being passed as a dict; it's being passed as a fully formed `B` class so the constructor of `A` should be able to use that information. Or is there a deeply baked reason that field values have to be converted to dict before parsing? If not I'm happy to dive in and send a PR if this one case is easy to fix.
